src/main/java/com/epam/esm/service/CertificateService.java

```diff
@@ -0,0 +1,205 @@
+package com.epam.esm.service;
+
+import static com.epam.esm.exceptions.Codes.CERTIFICATE_BAD_REQUEST;
+import static com.epam.esm.exceptions.Codes.CERTIFICATE_FOUND;
+import static com.epam.esm.exceptions.Codes.CERTIFICATE_NOT_FOUND;
+import static com.epam.esm.exceptions.Messages.CERTIFICATE_ALREADY_EXISTS;
+import static com.epam.esm.exceptions.Messages.CERTIFICATE_WITH_ID_NOT_FOUND;
+
+import com.epam.esm.Dto.Errors.ErrorDTO;
+import com.epam.esm.exceptions.CertificateNotFoundException;
+import com.epam.esm.model.GiftCertificate;
+import com.epam.esm.model.Tag;
+import com.epam.esm.repository.CertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.utils.CertificateSpecification;
+import jakarta.transaction.Transactional;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+public class CertificateService {
+
+    private final CertificateRepository certificateRepository;
+    private final TagRepository tagRepository;
+
+    @Autowired
+    public CertificateService(CertificateRepository certificateRepository, TagRepository tagRepository) {
+        this.certificateRepository = certificateRepository;
+        this.tagRepository = tagRepository;
+    }
+
+    public ResponseEntity<?> saveGiftCertificate(@NonNull GiftCertificate giftCertificate, List<Long> tagIdsList) {
+
+        Optional<ResponseEntity<ErrorDTO>> requestValidationMessage = validateCertificateRequest(giftCertificate, tagIdsList);
+        if (requestValidationMessage.isPresent()) return requestValidationMessage.get();
+
+        tagIdsList = tagIdsList.stream().distinct().collect(Collectors.toList());
```

Comment:
Here you have potential error: you need to do this before validation, otherwise validation would fail if you non-unique tag ids in the list.